### PR TITLE
Simplify structure of `MultiOptionQuestionDefinition` and remove subclasses

### PR DESCRIPTION
**Pull request by michaelzetune**
### Description

Currently the hierarchy for radio, checkbox, and dropdown questions is:

`Question` -> `MultiSelectQuestionDefinition` -> `RadioButtonQuestionDefinition`

and similarly for the other two. The only purpose of the three subclasses of `MultiSelectQuestionDefinition` is to specify that they should have the `SINGLE_SELECT_PREDICATE`. For example:

https://github.com/civiform/civiform/blob/7ce5f689d8f1f1b9d99717c9d448750dfa4b38b8/server/app/services/question/types/RadioButtonQuestionDefinition.java#L11-L55

There's some serious unreadable constructor bloat here, as well as in other `MultiSelectQuestionDefinition`s:

https://github.com/civiform/civiform/blob/7ce5f689d8f1f1b9d99717c9d448750dfa4b38b8/server/app/services/question/types/CheckboxQuestionDefinition.java#L11-L74

We can simplify this by just taking in a `MultiOptionQuestionType` in `MultiSelectQuestionDefinition` and removing the subclasses. 

In addition, for readability and the ability to remove lots of constructor overloads, we create a `MultiSelectQuestionDefinitionConfig` that allows us to abstract the construction of the configuration away from multiple constructors.

## Release notes

None.

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >
- [ ] Created tests which fail without the change (if possible)
   - New tests not necessary for refactor. Existing tests pass. 
- [ ] Extended the README / documentation, if necessary
   - Not necessary

#### User visible changes

None.

#### New Features

None.

### Issue(s) this completes

None.

### CODE REVIEW

File: server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      if (multiOptionQuestionType() == MultiOptionQuestionType.DROPDOWN
+          || multiOptionQuestionType() == MultiOptionQuestionType.CHECKBOX) {
+        // If we are using a dropdown or checkbox, set the SINGLE_SELECT_PREDICATE to ensure only
+        // one selection can be made.
+        setValidationPredicates(SINGLE_SELECT_PREDICATE);
+      }
+
+      return autoBuild();
+    }
+  }
```
Comment (michaelzetune) [inline server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java:87]: The `build()` method above replaces all the custom logic in `CheckboxQuestionDefinition` and `DropdownQuestionDefinition` in just a few lines; namely, that those types must have the `SINGLE_SELECT_PREDICATE`.

File: server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    return new AutoValue_MultiOptionQuestionDefinitionConfig.Builder();
+  }
+
+  static final MultiOptionValidationPredicates SINGLE_SELECT_PREDICATE =
+      MultiOptionValidationPredicates.create(1, 1);
+
+  public interface RequiredMultiOptionQuestionType {
+    Builder setMultiOptionQuestionType(MultiOptionQuestionType multiOptionQuestionType);
+  }
+
```
Comment (michaelzetune) [inline server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java:45]: The `RequiredMultiOptionQuestionType` interface simply forces the creator of a config to specify a question type, like:

```java
    MultiOptionQuestionDefinitionConfig config =
        MultiOptionQuestionDefinitionConfig.builder()
            .setMultiOptionQuestionType(MultiOptionQuestionType.CHECKBOX)
            .setName("name")
            .setEnumeratorId(Optional.empty())
            .setDescription("description")
            .setQuestionText(questionText)
            .setQuestionHelpText(questionHelpText)
            .setQuestionOptions(options)
            .setValidationPredicates(validationPredicates)
            .build();
```

File: server/app/controllers/dev/DatabaseSeedController.java [hunk 3]
```diff
@@ -141,22 +141,24 @@ private QuestionDefinition insertAddressQuestionDefinition() {
   }
 
   private QuestionDefinition insertCheckboxQuestionDefinition() {
-    return questionService
-        .create(
-            new CheckboxQuestionDefinition(
-                "checkbox",
-                Optional.empty(),
```
Comment (michaelzetune) [inline server/app/controllers/dev/DatabaseSeedController.java:148]: Here an elsewhere, this param `Optional.empty()` was the `enumeratorId`, and that is an optional field in the `MultiOptionQuestionDefinitionConfig` AutoValue, so it doesn't appear on the right side of the diff when set to empty.

File: server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.OptionalLong;
+import services.LocalizedStrings;
+import services.question.QuestionOption;
+import services.question.types.MultiOptionQuestionDefinition.MultiOptionValidationPredicates;
+
```
Comment (dkatzz) [inline server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java:11]: **suggestion:** Could you add a javadoc?

File: server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java (hunk 2 above)
Comment (dkatzz) [reply server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java:51]: Makes sense, thanks! Could you add a javadoc?

File: server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java (hunk 1 above)
Comment (dkatzz) [reply server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java:93]: So much better!!

**Review by dkatzz — APPROVED** [review]
Awesome job getting this cleaned up!

File: server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java (hunk 4 above)
Comment (michaelzetune) [reply server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java:11]: Done.

File: server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java (hunk 2 above)
Comment (michaelzetune) [reply server/app/services/question/types/MultiOptionQuestionDefinitionConfig.java:51]: Done.

**Pull request merged**